File: LeetCode/145二叉树的后序遍历.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> ans = new ArrayList<>();
        if (root == null) return ans;
        Stack<TreeNode> sta = new Stack<>();
        TreeNode pre = null;
        while (root != null || !sta.isEmpty()) {
            while (root != null) {
                sta.push(root);
                root = root.left;
            }
            root = sta.pop();
            if (root.right == null || root.right == pre) {
                ans.add(root.val);
                pre = root;
                root = null;
            } else {
                sta.push(root);
                root = root.right;
            }
        }
        return ans;
    }
}

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> ans = new ArrayList<>();
        Stack<TreeNode> sta = new Stack<>();
        if (root == null) return ans;
        sta.push(root);
        while (!sta.isEmpty()) {
            TreeNode cur = sta.pop();
            ans.add(cur.val);
            if (cur.left != null) sta.push(cur.left);
            if (cur.right != null) sta.push(cur.right);
        }
        Collections.reverse(ans);
        return ans;
    }
}

File: LeetCode/145二叉树的后序遍历.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func postorderTraversal(root *TreeNode) (ans []int) {
	sta := []*TreeNode{}
	var pre *TreeNode
	for root != nil || len(sta) > 0 {
		for root != nil {
			sta = append(sta, root)
			root = root.Left
		}
		root = sta[len(sta)-1]
		sta = sta[:len(sta)-1]
		if root.Right == nil || root.Right == pre {
			ans = append(ans, root.Val)
			pre = root
			root = nil
		} else {
			sta = append(sta, root)
			root = root.Right
		}
	}
	return
}